### DISCUSSION TITLE
What's the process for adding a new package/project?

### DISCUSSION THREAD

**Discussion by emadabdulrahim** [vercel/turborepo#243] [Help] [answered]
Say I want to add Remix project in the starter repo created by `npx create-turbo@latest`. What are the steps I need to take to have things wired up (including wiring the UI package) to the newly added Remix package?

Might be useful to add a section in the docs for that.

**Comment by kousherAlam** [comment]
I also want to know

**Comment by jaredpalmer** [comment, marked as answer]
Right now it's totally manual. You can copy and paste and then reinstall.

We know this isn't great, so code generation is something we are going to tackle early next year. 

In the meantime, I suggest using plop to make a template for yourself out of the packages or applications to meet your needs 

https://plopjs.com/documentation/#getting-started

**Reply by emadabdulrahim** [reply]
Thanks, that's what I've done. 

Since you showed an example of Remix project inside a turbo repo, I tried to do the same but seems like remix complains when I import a component from the `ui` package. Is that a remix specific isssue? Looks like an issue with supported import syntax. The error is "Unexpected export keyword ...." something along those lines.

```tsx
// inside remix file
import { Button } from 'ui'
```

**Reply by prichodko** [reply]
@emadabdulrahim this will be a Remix issue because it does not transpile its dependencies. You can see [here](https://github.com/vercel/turborepo/blob/main/examples/basic/apps/docs/next.config.js#L1) how Next.js works around this issue.

Modern bundlers like Vite and Parcel do this automatically.

**Reply by emadabdulrahim** [reply]
@prichodko Thanks for the tip

**Reply by nareshbhatia** [reply]
Lots of intricacies when trying to integrate remix with turborepo. I have overcome many of the issues by following the discussions here and in the remix repo (thanks to all the contributors). Here's my version of kitchen-sink that uses straight React, Next.js & Remix: https://github.com/nareshbhatia/custom-react-stack. See the issues section for Remix related issues.

**Reply by ricardo-fnd** [reply]
@jaredpalmer any news on when code generation will be available?

**Reply by nareshbhatia** [reply]
@jaredpalmer and everyone on this discussion,

I started looking at plop and other code generators in the last few months but was not satisfied with any. They were not monorepo friendly and did not make it easy to share generators. This prompted me to write my own. The goal was to make it modular and extensible. I have been able to hit both goals:

1. The generator is based on a plugin architecture which allows you to mix and match generators, giving you the ability to create a fully customized stack. For example, you can pick turborepo as your monorepo and react, nextjs or any other front-end framework as your base framework. Storybook, Jest & React Testing Library can be thrown in as options.
2. Writing a plugin is very easy. There is a plugin generator that scaffolds the basic infrastructure for you. You can make whatever tweaks you want (its all straight TypeScript) and add in your templates (EJS based). That's it - no rocket science.
3. Finally, the cli can load your plugins dynamically - just add them as devDependencies in the root package.json.

My hope is that we can build a strong ecosystem around this generator. To start off, I have created first-class documentation to get people started. More to come. I am looking for your feedback, support and most importantly contributions.

The name of this project is _Code Shaper_. The docs are published here: https://www.code-shaper.org/. Please let me know it this sounds interesting to you.

**Reply by yuval9313** [reply]
> Lots of intricacies when trying to integrate remix with turborepo. I have overcome many of the issues by following the discussions here and in the remix repo (thanks to all the contributors). Here's my version of kitchen-sink that uses straight React, Next.js & Remix: https://github.com/nareshbhatia/custom-react-stack. See the issues section for Remix related issues.

Can you add to the repo something to show how to use packages for the backend API?

**Reply by nareshbhatia** [reply]
> Can you add to the repo something to show how to use packages for the backend API?

Do you mean something like react-query on the client-side or express js on the server-side?

I have an example of using react-query on the client-side [here](https://github.com/code-shaper/movie-magic/blob/main/apps/movie-magic/src/pages/HomePage/useMovies.ts) with step-by-step instructions [here](https://www.code-shaper.dev/docs/getting-started/writing-a-custom-generator).

I will be creating a generator for an express js based API server as a Turborepo package in the next few days.

**Reply by nareshbhatia** [reply]
Hi @yuval9313, I now have a generator for creating an Express server in a Turborepo. You can find the docs for it here: https://www.code-shaper.dev/docs/reference/express-plugin

**Comment by miervaldis42** [comment]
Even if it is manually, you should **add a section in the docs** because I had a similar issue with adding a new app based on React, only to find out Turbo run `npm run dev` in each apps project, so you have to add it in your React script.
I am going to have the same wonders & questions when I am going to add a Node server. And actually, I understand a little more now why there are a lot of boilerplate code gists going around.
Probably, you will have a CLI with a command to add a new project to apps/ folder one day but think about devs who really want to use Turborepo in their project today but end up ditch it because of lack of documentation ^^"

**Reply by nareshbhatia** [reply]
@miervaldis42, our team went through the same pains initially and hence we created a scaffolding tool to solve this problem. You can start with a blank turborepo and add apps & libraries gradually - all the details are taken care of. So the CLI that you are looking for exists already! It is called Code Shaper. Please try it and give us your feedback, or even better - contribute!

https://code-shaper.dev/

**Comment by munafio** [comment]
for those who want add new app/package and work with a different frameworks at the same time using Turborepo, then you can take a look at [this starter example](https://github.com/munafio/react-vue-turborepo-starter) which is using (React and vue) at the same time with their `ui` packages.

**Comment by isomorpheric** [comment]
I'm having trouble adding existing projects to turborepo. People say just copy/paste, but what happens with the lockfile of the project being added? Do I delete it, and hope for no issues? Seems very wild west.

My dependencies keep getting messed up - projects that build fine outside of turbo, fail to build when adding them to turbo. This happens seemingly because of mismatching react versions, but I'm not even sure how.

I agree with the post above - this should be in the docs! I'm definitely rooting for TurboRepo, and Vercel is the right team to push it forward.

**Reply by isomorpheric** [reply]
I solved my problem and can confirm the issue was due to pnpm workspaces, my understanding of it and vscode being vscode (ctrl + p, reload window).

**Comment by engageintellect** [comment]
any updates here?

**Comment by ekafyi** [comment]
Posting for those who subscribe to this issue: https://turbo.build/repo/docs/core-concepts/monorepos/code-generation

**Comment by gunslingor** [comment]
I'm not understanding what turbo repo is... I thought it was a dynamic monorepo generator, but it can't setup and integrate express as a workspace? Even though many of templates have it integrated and even though many the commercial products depend on it? So confused... it feels like turborepo is just a set of 8 templates, not actually a monorepo manager... seems to just relly on manual workspaces for that, but that's node not turborepo right? So what's turborepo? Spent 8 hours studying it, I got further trying to setup express using intellij templates and ground up... but alas, vercel was a brick wall there, 8 hours wasted trying to deploy hello world express from a monorepo. I think my stint into vercel products might be coming to an end, gcloud seems much easier.

I expected to be able to go:
Turbo gen blank
Turbo workspace express backend1 backend2
Turbo workspace vite frontend
Turbo workspace react-native native
Turbo workspace typescript shared_code
Turbo workspace typescript database

Something like that... all I see are templates in this product? What am I missing?
Turbo

**Reply by nareshbhatia** [reply]
@gunslingor, you can think of turborepo as a build tool that allows you to manage the workspaces in a monorepo. What the content of each workspace is – that's your responsibility, e.g. an express backend, a react frontend etc. If you want you create such workspaces easily, look at [Code Shaper](https://www.code-shaper.dev/) – it allows you to generate different kinds of workspaces on top of turborepo. Go through the Getting Started section to understand how it works.
